src/features/device-logs/lib/read.ts

```diff
@@ -144,16 +144,16 @@ async function handleStreamingRead(
 
 		const afterDate = logs.length && logs[logs.length - 1].createdAt;
 		// Append the subscription logs to the history queue
-		while (buffer.length) {
-			const log = buffer.shift();
+		for (const log of buffer) {
 			if (log && log.createdAt > afterDate) {
```

Comment:
Presumably these need to be in date order, in which case can we find the index of the first valid log and then add all of those to the `logs` array in one go instead of continuing to iterate?

Reply:
@Page- you means with something like v ?
```js
		const firstAfterDateIndex = buffer.findIndex(log => log.createdAt > afterDate);
		if (firstAfterDateIndex > 0) {
			buffer.splice(0, firstAfterDateIndex);
		}
		if (firstAfterDateIndex !== -1 && buffer.length > 0) {
			logs = logs.concat(buffer);
		}
```
Using `,concat` seems to be both faster than `.push(...buffer)` and safer than `.splice(0, 0, ...buffer)` (which for length > 200_000 starts throwing an `RangeError: Maximum call stack size exceeded`).
I didn't account for the GC that `.concat` will cause later though.

Reply:
Yes that is what I meant (via any of the methods you mentioned)